### PR TITLE
feat(dataflow): Add Go Flex Template example

### PR DESCRIPTION
**Pull request by jrmccluskey**
Adds a code snippet and instructions on how to build and run a Dataflow flex template using the Apache Beam Go SDK. This is meant to be the Go equivalent of the [Python](https://github.com/GoogleCloudPlatform/python-docs-samples/tree/main/dataflow/flex-templates/streaming_beam) and [Java](https://github.com/GoogleCloudPlatform/java-docs-samples/tree/main/dataflow/flex-templates/streaming_beam_sql) tutorials, albeit with a simpler wordcount pipeline.

### CODE REVIEW

File: dataflow/flex-templates/wordcount/metadata.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+            "isOptional": true
+        },
+        {
+            "name": "output",
+            "label": "Output file.",
+            "helpText": "GCS location to write output to.",
+            "isOptional": false
+        }
+    ]
+}
```
Comment (enocom) [inline dataflow/flex-templates/wordcount/metadata.json:18]: nit: newline

File: dataflow/flex-templates/wordcount/wordcount.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+// formatFn is a DoFn that formats a word and its count as a string.
+func formatFn(w string, c int) string {
+	return fmt.Sprintf("%s: %v", w, c)
+}
+
+// CountWords is a composite transform that counts the words of a PCollection
+// of lines. It expects a PCollection of type string and returns a PCollection
+// of type KV<string,int>. The Beam type checker enforces these constraints
+// during pipeline construction.
+func CountWords(s beam.Scope, lines beam.PCollection) beam.PCollection {
```
Comment (enocom) [inline dataflow/flex-templates/wordcount/wordcount.go:89]: Does this function need to be exported?

File: dataflow/flex-templates/wordcount/wordcount.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	smallWords      = beam.NewCounter("extract", "smallWords")
+	lineLen         = beam.NewDistribution("extract", "lineLenDistro")
+)
+
+// extractFn is a DoFn that emits the words in a given line and keeps a count for small words.
+type extractFn struct {
+	SmallWordLength int `json:"smallWordLength"`
+}
+
+func (f *extractFn) ProcessElement(ctx context.Context, line string, emit func(string)) {
```
Comment (enocom) [inline dataflow/flex-templates/wordcount/wordcount.go:65]: I assume the Beam Pipeline requires this to be exported. Is that right?

Also, a GoDoc for the method would be nice.

**Review by enocom — CHANGES_REQUESTED** [review]
Should this have some tests?

File: dataflow/flex-templates/wordcount/wordcount.go (hunk 3 above)
Comment (jrmccluskey) [reply dataflow/flex-templates/wordcount/wordcount.go:67]: Yes the lifecycle methods for structure DoFns have to be exported, along with any internal fields that need to be serialized (like SmallWordLength.)

File: dataflow/flex-templates/wordcount/metadata.json (hunk 1 above)
Comment (jrmccluskey) [reply dataflow/flex-templates/wordcount/metadata.json]: Done.

File: dataflow/flex-templates/wordcount/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,10 @@
+FROM gcr.io/dataflow-templates-base/go-template-launcher-base
+
+ARG WORKDIR=/dataflow/template
+RUN mkdir -p ${WORKDIR}
+
+COPY wordcount ${WORKDIR}/wordcount
+
+ENV FLEX_TEMPLATE_GO_BINARY="${WORKDIR}/wordcount"
+
+ENTRYPOINT ["/opt/google/dataflow/go_template_launcher"]
```
[the remark this answers is not in the document]
Comment (jrmccluskey) [reply dataflow/flex-templates/wordcount/Dockerfile]: Done.

File: dataflow/flex-templates/wordcount/wordcount.go (hunk 2 above)
Comment (jrmccluskey) [reply dataflow/flex-templates/wordcount/wordcount.go:91]: I don't think it has to be, it could be made unexported if you would prefer.

**Comment by jrmccluskey** [comment]
We have [some tests for this that exist already](https://github.com/apache/beam/blob/7fea1ef262cad725ce30651ebb441d56dba0c11c/sdks/go/test/integration/wordcount/wordcount_test.go), although they're somewhat abstract since they leverage the counters and use some hashing to validate the output instead of looking at the PCollection content. I can add them if you'd prefer a test example with the snippet though.

**Comment by tritone** [comment]
> We have [some tests for this that exist already](https://github.com/apache/beam/blob/7fea1ef262cad725ce30651ebb441d56dba0c11c/sdks/go/test/integration/wordcount/wordcount_test.go), although they're somewhat abstract since they leverage the counters and use some hashing to validate the output instead of looking at the PCollection content. I can add them if you'd prefer a test example with the snippet though.

We require integration tests for all samples in this repo -- see https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md#testing for guidelines.

File: dataflow/flex-templates/wordcount/wordcount.go (hunk 2 above)
Comment (enocom) [reply dataflow/flex-templates/wordcount/wordcount.go:91]: In that case, not exporing from package main is preferred.

**Comment by enocom** [comment]
Some kind of test coverage would be good, if only to ensure this sample keeps working for the future.

**Comment by jrmccluskey** [comment]
Added tests

**Review by enocom — APPROVED** [review]

**Pull request merged**